File: Java/BoundService/app/src/main/java/com/abdelrahman/java/boundservice/MainActivity.java
package com.abdelrahman.java.boundservice;

import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.os.Handler;
import android.os.IBinder;
import android.util.Log;

import com.abdelrahman.java.boundservice.databinding.ActivityMainBinding;

public class MainActivity extends AppCompatActivity {
    private ActivityMainBinding binding;
    private OdometerService odometerService;
    private boolean isBounded = false;

    private ServiceConnection connection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName componentName, IBinder binder) {
            OdometerService.OdometerBinder odometerBinder = (OdometerService.OdometerBinder) binder;
            odometerService = odometerBinder.getOdometer();
            isBounded = true;
        }

        @Override
        public void onServiceDisconnected(ComponentName componentName) {
            isBounded = false;
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main);
        watchDistanceChange();
    }

    @Override
    protected void onStart() {
        super.onStart();
        Intent intent = new Intent(this,OdometerService.class);
        bindService(intent,connection, Context.BIND_AUTO_CREATE);
        /**
         * The code Context.BIND_AUTO_CREATE tells Android to
         * create the service if it doesn’t already exist.
         */
    }

    private void watchDistanceChange() {
        final Handler handler = new Handler();
        handler.post(new Runnable() {
            @Override
            public void run() {
                double distance = 0.0;
                if(odometerService != null){
                    distance = Math.round(odometerService.getMeters() * 100.0) / 100.0;
                }
                binding.distance.setText(distance+" Meter");
                handler.postDelayed(this,1000);
            }
        });
    }



    @Override
    protected void onStop() {
        super.onStop();
        if(isBounded){
            unbindService(connection);
            isBounded=false;
        }
    }
}

File: Kotlin/BoundService/app/src/main/java/com/abdelrahman/koltin/boundservice/MainActivity.kt
package com.abdelrahman.koltin.boundservice

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.os.Handler
import android.os.IBinder
import android.util.Log
import androidx.core.os.postDelayed
import androidx.databinding.DataBindingUtil
import com.abdelrahman.koltin.boundservice.databinding.ActivityMainBinding
import kotlinx.android.synthetic.main.activity_main.*
import kotlin.math.roundToInt

class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding
    private lateinit var odometerService: OdometerService
    private var isBound = false

    private val connection = object : ServiceConnection {
        override fun onServiceDisconnected(p0: ComponentName?) {
            isBound = false
        }

        override fun onServiceConnected(p0: ComponentName?, binder: IBinder?) {
            val odometerBinder = binder as OdometerService.OdometerBinder
            odometerService = odometerBinder.getOdometer()
        }

    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main)
        watchDistanceChange()
    }

    override fun onStart() {
        super.onStart()
        Log.d("MainActivity", "Start")
        bindService(Intent(this, OdometerService::class.java), connection, Context.BIND_AUTO_CREATE)
        /**
         * The code Context.BIND_AUTO_CREATE tells Android to
         * create the service if it doesn’t already exist.
         */
    }

    private fun watchDistanceChange() {
        val handler = Handler()
        handler.postDelayed(object : Runnable {
            override fun run() {
                Log.d("MainActivity", "run")
                var distance = 0.0
                if (odometerService != null)
                    distance = (odometerService.getMeters() * 100.0).roundToInt() / 100.00
                val stringDistance: String = "%.2f".format(distance)
                binding.distance.text = stringDistance
                binding.distance.append("Meter")
                handler.postDelayed(this, 1000)
            }

        }, 1000)
    }

    override fun onStop() {
        Log.d("MainActivity", "Stopped")
        super.onStop()
        if (isBound) {
            unbindService(connection)
            isBound = false
        }
    }
}
